src/shogun/clustering/KMeans.cpp

```diff
@@ -331,57 +339,84 @@ void CKMeans::store_model_features()
 
 SGMatrix<float64_t> CKMeans::kmeanspp()
 {
-	int32_t num=distance->get_num_vec_lhs();
-	SGVector<float64_t> dists=SGVector<float64_t>(num);
-	SGVector<int32_t> mu_index=SGVector<int32_t>(k);
-
-	/* 1st center */
-	int32_t mu_1=CMath::random((int32_t) 0,num-1);
-	mu_index[0]=mu_1;
-
-	/* choose a center - do k-1 times */
-	int32_t count=0;
-	while (++count<k)
-	{
-		float64_t sum=0.0;
-		/* for each data point find distance to nearest already chosen center */
-		for (int32_t point_idx=0;point_idx<num;point_idx++)
+	int32_t lhs_size;
+	CDenseFeatures<float64_t>* lhs=(CDenseFeatures<float64_t>*)distance->get_lhs();
+	lhs_size=lhs->get_num_vectors();
+	
+	SGMatrix<float64_t> centers=SGMatrix<float64_t>(dimensions, k);
+	centers.zero();
+	SGVector<float64_t> min_dist=SGVector<float64_t>(lhs_size);
+	min_dist.zero();
+
+	/* First center is chosen at random */
+	int32_t mu=CMath::random((int32_t) 0, lhs_size-1);
+	SGVector<float64_t> mu_first=lhs->get_feature_vector(mu);	
+	for(int32_t j=0; j<dimensions; j++)
+		centers(j, 0)=mu_first[j];
+
+	distance->precompute_lhs();
```

Comment:
I think this name can lead to confusion because it is not distances that are pre-computed but the squared norms of the vectors. Why did the name change in #3073, @Saurabh7?